File: Leetcode Daily Challenge/017-November-2024/002-circular-sentence.cpp
// https://leetcode.com/problems/circular-sentence/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(N)
    // SC : O(N) At worst
    bool isCircularSentence(string sentence) {
        int n = sentence.length();
        if (sentence[0] != sentence[n - 1]) 
            return false;

        istringstream ss(sentence);
        string word;
        char prev = '$';

        while (getline(ss, word, ' ')) { // while (ss >> word) : this sytax can also be used
            int len = word.length();
            if (prev == '$') 
                prev = word[len - 1];
            else {
                if (prev != word[0])
                    return false;
                else 
                    prev = word[len - 1];
            }
        }

        return true;
    }
};

File: Leetcode Daily Challenge/017-November-2024/002-circular-sentence.java
// https://leetcode.com/problems/circular-sentence/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public boolean isCircularSentence(String sentence) {
        int n = sentence.length();
        if (sentence.charAt(0) != sentence.charAt(n - 1))
            return false;

        Scanner scanner = new Scanner(sentence);
        char prev = '$';

        while (scanner.hasNext()) {
            String word = scanner.next();
            int len = word.length();
            if (prev == '$') {
                prev = word.charAt(len - 1);
            } else {
                if (prev != word.charAt(0)) {
                    return false;
                } else {
                    prev = word.charAt(len - 1);
                }
            }
        }   

        scanner.close();
        return true;
    }
}